runhouse/servers/env_servlet.py

```diff
@@ -227,61 +254,100 @@ def _get_env_cpu_usage(self, cluster_config: dict = None):
-                    "total_memory": total_gpu_memory,  # in bytes
-                }
-            else:
-                env_gpu_usage = {}
-        except subprocess.CalledProcessError as e:
-            logger.error(f"Failed to get GPU usage for {self.env_name}: {e}")
-            env_gpu_usage = {}
+            total_used_memory = sum_used_memery / len(collected_gpu_info)  # average
+
+        total_used_memory = total_used_memory / len(gpus)
+
+        return {"total_memory": total_gpu_memory, "used_memory": total_used_memory}
+
+    def _collect_env_gpu_usage(self):
+        """periodically collects env gpu usage"""
+
+        pynvml.nvmlInit()  # init nvidia ml info collection
+
+        cluster_config = obj_store.cluster_config
+        interval_size = cluster_config.get(
+            "status_check_interval", DEFAULT_STATUS_CHECK_INTERVAL
+        )
 
-        return env_gpu_usage
+        while True:
+            try:
+                # TODO: [sb]: once introduced, we could use ClusterServlet _cluster_periodic_thread_alive() to replace the
+                #  'should_send_status_and_logs_to_den' logic below.
+                # Only if one of these is true, do we actually need to get the status from each EnvServlet
+                should_send_status_and_logs_to_den: bool = (
+                    configs.token is not None and interval_size != -1
+                )
+                if not should_send_status_and_logs_to_den:
+                    pynvml.nvmlShutdown()
+                    break
+
+                gpu_count = pynvml.nvmlDeviceGetCount()
+                with self.lock:
+                    if not self.gpu_info:
+                        self.gpu_info = {device: [] for device in range(gpu_count)}
+
+                    for gpu_index in range(gpu_count):
+                        handle = pynvml.nvmlDeviceGetHandleByIndex(gpu_index)
+                        processes = pynvml.nvmlDeviceGetComputeRunningProcesses_v3(
+                            handle
+                        )
+                        memory_info = pynvml.nvmlDeviceGetMemoryInfo(handle)
+                        if processes:
+                            for p in processes:
+                                if p.pid == self.pid:
+                                    used_memory = p.usedGpuMemory  # in bytes
+                                    total_memory = memory_info.total  # in bytes
+
+                                    updated_gpu_info = self.gpu_info[gpu_index]
+                                    # to reduce cluster memory usage (we are saving the gpu_usage info on the cluster),
+                                    # we save only the most updated gpu usage. If for some reason the size of updated_gpu_info is
+                                    # too big, we remove the older gpu usage info.
+                                    if len(updated_gpu_info) + 1 > MAX_GPU_INFO_LEN:
+                                        cut_a_third = int(
```

Comment:
Is this third arbitrary? Could be worth adding explanation for the choice or maybe making it a variables/constant

Reply:
Yeah, it's a good point. I've added it as a constant, along with some explanation for its value choice.